File: MatlabCodes/Fig03_07.m
function BS_Hedging
clf;clc;close all;
format long

% Model configuration

S0         = 1;
NoOfPaths = 1000;
NoOfSteps = 1000;
r         = 0.1;
sigma     = 0.2;
T         = 1;
K         = 1*0.95;

% Generation of the stock paths

randn('seed',11) %17 %20 %25 %43 %4 %11
[S,time] = GeneratePathsGBMEuler(NoOfPaths,NoOfSteps,T,r,sigma,S0);

% Define lambda expression for call option and for delta

C = @(t,K,S0)BS_Call_Put_Option_Price('c',S0,K,sigma,t,T,r);
Delta_call = @(t,K,S0)BS_Delta('c',S0,K,sigma,t,T,r);

% Hedging part

PnL        = zeros(NoOfPaths,length(time));
delta_init = Delta_call(0,K,S0);
PnL(:,1)   = C(0,K,S0) - delta_init*S0;

CallM      = zeros(NoOfPaths,length(time));
CallM(:,1) = C(0,K,S0);
DeltaM     = zeros(NoOfPaths,length(time));
DeltaM(:,1)= Delta_call(0,K,S0);

% Iterate over all the time steps

for i=2:NoOfSteps+1
    time(i)
    dt = time(i)-time(i-1);   
    delta_old  = Delta_call(time(i-1),K,S(:,i-1));
    delta_curr = Delta_call(time(i),K,S(:,i));
    PnL(:,i)   =  PnL(:,i-1)*exp(r*dt) - (delta_curr-delta_old).*S(:,i); % PnL
    CallM(:,i) = C(time(i),K,S(:,i));
    DeltaM(:,i)= delta_curr;
end

% Final PnL calculation 

PnL(:,end)=PnL(:,end) -max(S(:,end)-K,0) +  DeltaM(:,end).*S(:,end);

% Plot hedging results for a specified path

path_id =21;
 
figure(1)
plot(time,S(path_id,:),'linewidth',1.5)
hold on
plot(time,CallM(path_id,:),'-m','linewidth',1.5)
plot(time,PnL(path_id,:),'-r','linewidth',1.5)
plot(time,DeltaM(path_id,:),'g','linewidth',1.5)
plot(time,K.*ones(length(time),1),'--k','linewidth',1.5)
legend('Stock value','Call price','P&L','\Delta')
grid on
grid minor
title('Hedging of a Call opton')
xlabel('time')
xlim([0,T+0.01])

% PnL histogram

figure(2)
hist(PnL(:,end),50)
xlim([-0.1,0.1])
grid on
title('PnL Histogram with respect to number of re-hedging')

% Analysis for each path

for i =1:NoOfPaths
    fprintf('path_id = %2.0d, PnL(t_0)= %0.4f, PnL(Tm-1) =%0.4f,S(Tm) =%0.4f,max(S(tm)-K,0)= %0.4f, PnL(t_m) = %0.4f',i,PnL(i,1),PnL(i,end-1),S(i,end),max(S(i,end)-K,0),PnL(i,end))
    fprintf('\n')
end

function [S,time] = GeneratePathsGBMEuler(NoOfPaths,NoOfSteps,T,r,sigma,S0)

% Approximation

S=zeros(NoOfPaths,NoOfSteps);
S(:,1) = S0;

% Random noise

Z=random('normal',0,1,[NoOfPaths,NoOfSteps]);
W=zeros([NoOfPaths,NoOfSteps]);

dt = T / NoOfSteps;
time = zeros([NoOfSteps+1,1]);
for i=1:NoOfSteps
    if NoOfPaths>1
        Z(:,i)   = (Z(:,i) - mean(Z(:,i))) / std(Z(:,i));
    end
    W(:,i+1)  = W(:,i) + sqrt(dt).*Z(:,i);
    S(:,i+1) = S(:,i) + r * S(:,i)*dt + sigma * S(:,i).*(W(:,i+1)-W(:,i));
    time(i+1) = time(i) + dt;
end

function value = BS_Delta(CP,S_0,K,sigma,t,T,r)
if (t-T>10e-20 && T-t<10e-7)
   t=T;
end
d1    = (log(S_0 ./ K) + (r + 0.5 * sigma.^2) * (T-t)) ./ (sigma * sqrt(T-t));
if lower(CP) == 'c' || lower(CP) == 1
   value = normcdf(d1);
elseif lower(CP) == 'p' || lower(CP) == -1
    value = normcdf(d1)-1;
end

% Black-Scholes call option price

function value = BS_Call_Put_Option_Price(CP,S_0,K,sigma,t,T,r)
if (t-T>10e-20 && T-t<10e-7)
   t=T;
end
d1    = (log(S_0 ./ K) + (r + 0.5 * sigma.^2) * (T-t)) ./ (sigma * sqrt(T-t));
d2    = d1 - sigma * sqrt(T-t);
if lower(CP) == 'c' || lower(CP) == 1
    value = normcdf(d1) .* S_0 - normcdf(d2) .* K * exp(-r * (T-t));
elseif lower(CP) == 'p' || lower(CP) == -1
    value = normcdf(-d2) .* K*exp(-r*(T-t)) - normcdf(-d1).*S_0;
end

File: PythonCodes/Chapter 03/Fig03_07.py
#%%
"""
Created on Thu Dec 12 2018
Hedging with the Black Scholes model
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st
import enum 
from mpl_toolkits import mplot3d
from scipy.interpolate import RegularGridInterpolator

# This class defines puts and calls

class OptionType(enum.Enum):
    CALL = 1.0
    PUT = -1.0

def GeneratePathsGBM(NoOfPaths,NoOfSteps,T,r,sigma,S_0):    
    Z = np.random.normal(0.0,1.0,[NoOfPaths,NoOfSteps])
    X = np.zeros([NoOfPaths, NoOfSteps+1])
    W = np.zeros([NoOfPaths, NoOfSteps+1])
    time = np.zeros([NoOfSteps+1])
        
    X[:,0] = np.log(S_0)
    
    dt = T / float(NoOfSteps)
    for i in range(0,NoOfSteps):

    # Making sure that samples from the normal distribution have mean 0 and variance 1

        if NoOfPaths > 1:
            Z[:,i] = (Z[:,i] - np.mean(Z[:,i])) / np.std(Z[:,i])
        W[:,i+1] = W[:,i] + np.power(dt, 0.5)*Z[:,i]
        X[:,i+1] = X[:,i] + (r - 0.5 * sigma * sigma) * dt + sigma * (W[:,i+1]-W[:,i])
        time[i+1] = time[i] +dt
        
    # Compute exponent of ABM

    S = np.exp(X)
    paths = {"time":time,"S":S}
    return paths

# Black-Scholes call option price

def BS_Call_Put_Option_Price(CP,S_0,K,sigma,t,T,r):
    K = np.array(K).reshape([len(K),1])
    d1    = (np.log(S_0 / K) + (r + 0.5 * np.power(sigma,2.0)) 
    * (T-t)) / (sigma * np.sqrt(T-t))
    d2    = d1 - sigma * np.sqrt(T-t)
    if CP == OptionType.CALL:
        value = st.norm.cdf(d1) * S_0 - st.norm.cdf(d2) * K * np.exp(-r * (T-t))
    elif CP == OptionType.PUT:
        value = st.norm.cdf(-d2) * K * np.exp(-r * (T-t)) - st.norm.cdf(-d1)*S_0
    return value

def BS_Delta(CP,S_0,K,sigma,t,T,r):
    # When defining a time grid it may happen that the last grid point 
    # lies slightly a bit behind the maturity time

    if t-T>10e-20 and T-t<10e-7:
        t=T
    K = np.array(K).reshape([len(K),1])
    d1    = (np.log(S_0 / K) + (r + 0.5 * np.power(sigma,2.0)) * \
             (T-t)) / (sigma * np.sqrt(T-t))
    if CP == OptionType.CALL:
        value = st.norm.cdf(d1)
    elif CP == OptionType.PUT:
       value = st.norm.cdf(d1)-1.0
    return value


def mainCalculation():
    NoOfPaths = 1000
    NoOfSteps = 200
    T         = 1.0
    r         = 0.1
    sigma     = 0.2
    s0        = 1.0
    K         = [0.95]
    CP        = OptionType.CALL
    
    np.random.seed(1)
    Paths = GeneratePathsGBM(NoOfPaths,NoOfSteps,T,r,sigma,s0)
    time  = Paths["time"]
    S     = Paths["S"]
    
    # Setting up some handy lambda values

    C = lambda t,K,S0: BS_Call_Put_Option_Price(CP,S0,K,sigma,t,T,r)
    Delta = lambda t,K,S0: BS_Delta(CP,S0,K,sigma,t,T,r)
    
    # Setting up the initial portfolio

    PnL = np.zeros([NoOfPaths,NoOfSteps+1])
    delta_init= Delta(0.0,K,s0)
    PnL[:,0] = C(0.0,K,s0) - delta_init * s0
            
    CallM      = np.zeros([NoOfPaths,NoOfSteps+1])
    CallM[:,0] = C(0.0,K,s0)
    DeltaM     = np.zeros([NoOfPaths,NoOfSteps+1])
    DeltaM[:,0] = Delta(0,K,s0)
    
    for i in range(1,NoOfSteps+1):
        dt = time[i] - time[i-1]
        delta_old  = Delta(time[i-1],K,S[:,i-1])
        delta_curr = Delta(time[i],K,S[:,i])
        
        PnL[:,i]    =  PnL[:,i-1]*np.exp(r*dt) - (delta_curr-delta_old)*S[:,i] # PnL
        CallM[:,i]  = C(time[i],K,S[:,i])
        DeltaM[:,i] = delta_curr
    
    # Final transaction, payment of the option (if in the money) and selling the hedge

    PnL[:,-1] = PnL[:,-1] -np.maximum(S[:,-1]-K,0) +  DeltaM[:,-1]*S[:,-1]
    
    # We plot only one path at the time

    path_id = 13
    plt.figure(1)
    plt.plot(time,S[path_id,:])
    plt.plot(time,CallM[path_id,:])
    plt.plot(time,DeltaM[path_id,:])
    plt.plot(time,PnL[path_id,:])
    plt.legend(['Stock','CallPrice','Delta','PnL'])
    plt.grid()
    
    # Plot the histogram of the PnL

    plt.figure(2)
    plt.hist(PnL[:,-1],50)
    plt.grid()
    plt.xlim([-0.1,0.1])
    plt.title('histogram of P&L')
    
    # Analysis for each path

    for i in range(0,NoOfPaths):
        print('path_id = {0:2d}, PnL(t_0)={1:0.4f}, PnL(Tm-1) ={2:0.4f},S(t_m) = {3:0.4f}, max(S(tm)-K,0)= {4:0.4f}, PnL(t_m) = {5:0.4f}'.format(i,PnL[0,0],
              PnL[i,-2],S[i,-1],np.max(S[i,-1]-K,0),PnL[i,-1]))
mainCalculation()
